[#596] Initial implementation of `SCYLLA_USE_METADATA_ID` protocol feature negotiation

nikagra opened the pull request:
# Issue [java-driver#596](https://github.com/scylladb/java-driver/issues/596)

repared statements invalidation used to be completely broken, you can read details on it [here](https://github.com/scylladb/scylladb/issues/20860)
Recently core merged [PR](https://github.com/scylladb/scylladb/pull/23292) that fixes problems for select statements.
After that PR, if driver supply `SCYLLA_USE_METADATA_ID` at the startup, when statements is prepared server hands out result queries metadata hash.
When driver executes this query it sends metadata id alongside with the request.
Server checks driver metadata id and what it has on its side, if there is any difference it will send new metadata and metadata id.
Driver suppose to pick it up and update metadata and metadata id on it's side, before deserializing response, solving issue with having outdated metadata on driver side.
Unfortunately it is solved only for SELECT statements, UPDATE/INSERT will be solved on separate occasion.

Core changes are in, we need to start working on driver side.

## To Be Done:

1. parse `SCYLLA_USE_METADATA_ID` protocol extension from SUPPORTED frame;
    - if present in SUPPORTED, send it in the STARTUP frame;
2. if the extension is negotiated:
    - store the metadata ID, same way it is done for CQL v5
    - read `METADATA_CHANGED` (`=3`) flag in result metadata;
    - if the flag is set, read the new metadata ID (short bytes) and update it on prepare statement
3. if received new result metadata that is different from the cached one:
    - use the new metadata for deserialization instead of the old one;
    - replace the old cached metadata with the new one.

## Resolve skip metadata flag
Default value of skip metadata flag should be treated as `safest option`
Which means that if  `SCYLLA_USE_METADATA_ID` was negotiated or CQL v5 is used then it is true, result metadata is skipped.
In other cases it is false.

# Changes
* Negotiating `SCYLLA_USE_METADATA_ID` extension during protocol initialization `Options` step.
* Refactoring `ProtocolFeature` enum to `ProtocolFeatures` class enhancing `isSUpportedBy` method.
* Introducing and storing in the channel `ProtocolFeatureStore` instance, which stores information about optional protocol features.
* Enhancing `Message` to load `ProtocolFeatureStore` from channel and to later use it for encoding and decoding (in encoders and decoders).

nikagra committed:
[#596] Implementation of `SCYLLA_USE_METADATA_ID` protocol feature negotiation. Introducing `ProtocolFeatureStore`.

[#596] Initial implementation of SCYLLA_USE_METADATA_ID feature.

[#596] Fixing NPE when accessing sharding info from empty ProtocolFeatureStore.

[#596] Storing `ProtocolFeatureStore` in `ProtocolEncoder`/`ProtocolDecoder`. Getting read of sharing it via `Channel` attributes.

[#596] Addressing missing `ProtocolFeatureStore` in a `statement` param of `executeAsync` to fix ITs.

[#596] Adjusting the way feature store is handled in `DefaultResultSetFuture`. Removing feature store from its constructor.

[#596] Moving channel attribute management for feature store to utility methods of `ProtocolFeatureStore`.

 [#596] Setting `Host`'s feature store separately from `Channel`. Code clean up

Apply suggestion from @dkropachev

Co-authored-by: Dmitry Kropachev <dmitry.kropachev@gmail.com>

 [#596] `ProtocolEncoder` clean up.

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Host.java:
```diff
@@ -621,8 +634,7 @@ public interface StateListener {
     /**
      * Gets invoked when the listener is registered with a cluster, or at cluster startup if the
-     * listener was registered at initialization with {@link
-     * com.datastax.driver.core.Cluster#register(Host.StateListener)}.
```
Please move these kind of changes into a separate commit with proper message to it that explains why it is done

nikagra replied:
This should've been some automated change, but what tool has performed it. I will revert the changes

dkropachev replied:
I still see them, please remove them from the PR.

dkropachev replied:
Still see them

nikagra committed:
[#596] Improving Scylla test coverage in `PreparedStatementInvalidationTest`. Registering for closure resources implementing `AutoCloseable` interface in flaky tests class.

[#596] Improving Scylla test coverage in `PreparedStatementInvalidationTest`.

Apply suggestion from @dkropachev

Rearranging assertions and minor improvements in `should_never_update_statement_id_for_conditional_updates`

Co-authored-by: Dmitry Kropachev <dmitry.kropachev@gmail.com>

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -449,35 +450,31 @@ private AsyncFunction<Message.Response, Void> onOptionsResponse(
-      public ListenableFuture<Void> apply(Message.Response response) throws Exception {
-        switch (response.type) {
+      public ListenableFuture<Void> apply(Message.Response response) {
+        switch (Objects.requireNonNull(response).type) {
```
why do you need to check that ?

nikagra replied:
At some point I've replaced anonymous class with a lambda, a warning about `response` nullability appeared after, which I addressed. I've later reverted back to anonymous class to reduce number of optional changes. I will revert this change as well

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/DefaultResultSetFuture.java:
```diff
@@ -87,15 +87,21 @@ public void onSet(
                     rm.getCustomPayload().get(TabletInfo.TABLETS_ROUTING_V1_CUSTOM_PAYLOAD_KEY));
           }
+          ProtocolFeatureStore featureStore =
+              ProtocolFeatureStore.loadFromChannel(connection.channel);
```
it should have been defined when `DefaultResultSetFeature` is created, move it to the class

nikagra replied:
Do you mean passing `protocolFeatureStore` as a constructor argument? This would mean that fetching feature store from channel will be happening before constructor call. I don't think that performing this type of task within constructor is a good idea.

dkropachev replied:
sorry, i wasn't clear, make `ProtocolFeatureStore` a constructor arugment.

dkropachev replied:
sorry, i wasn't clear, make `ProtocolFeatureStore` a constructor arugment.

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Message.java:
```diff
@@ -318,10 +325,12 @@ static class ProtocolEncoder extends MessageToMessageEncoder<Request> {
 
     @Override
     protected void encode(ChannelHandlerContext ctx, Request request, List<Object> out) {
+      ProtocolFeatureStore featureStore = ProtocolFeatureStore.loadFromChannel(ctx.channel());
```
Update `ProtocolEncoder` to store `ProtocolFeatureStore`, `messageEncoderFor` to construct `ProtocolEncoder` on flight instead of silly static picker they currently have.
And update `switchToModernFraming` to rebuild pipeline with new encoder.

nikagra commented on driver-core/src/main/java/com/datastax/driver/core/SessionManager.java:
```diff
@@ -141,10 +141,20 @@ public String getLoggedKeyspace() {
+            .map(Host::getProtocolFeatureStore)
+            .orElse(
+                ProtocolFeatureStore
+                    .EMPTY); // Assuming feature stores are the same, and we can use any of them
```
@dkropachev Do you think this way of fetching feature store is ok here?

dkropachev replied:
just pushed commit with changes to address it, please take a look

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Host.java:
```diff
@@ -80,25 +73,27 @@ enum State {
   private volatile Set<Token> tokens;
 
   private volatile String dseWorkload;
+
```
Could you please remove these changes

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Host.java:
```diff
@@ -437,29 +432,39 @@ void setTokens(Set<Token> tokens) {
 
-  public void setShardingInfo(ShardingInfo shardingInfo) {
-    this.shardingInfo = shardingInfo;
+  public void setProtocolFeatureStore(ProtocolFeatureStore protocolFeatureStore) {
```
Can we make it protected ? it is definitely not for public use

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/ProtocolFeatureStore.java:
```diff
@@ -0,0 +1,98 @@
+  public void storeInChannelAndHost(Channel channel, Host host) {
+    host.setProtocolFeatureStore(this);
+    channel.attr(ProtocolFeatureStore.CHANNEL_KEY).set(this);
+  }
```
Don't couple like this, there is no reason to keep them together, move `setProtocolFeatureStore` out please.

dkropachev commented on driver-core/src/test/java/com/datastax/driver/core/PreparedStatementInvalidationTest.java:
```diff
@@ -182,21 +236,44 @@ public void should_not_reprepare_invalid_statements() {
     // Given
+    boolean isScylla = Objects.nonNull(ccm().getScyllaVersion());
+    boolean supportsUseMetadataId =
+        isScylla && ccm().getScyllaVersion().compareTo(SCYLLA_METADATA_ID_SUPPORT_VERSION) > 0;
```
```suggestion
    boolean supportsUseMetadataId = isSupportsUseMetadataId();
```

dkropachev commented on driver-core/src/test/java/com/datastax/driver/core/PreparedStatementInvalidationTest.java:
```diff
@@ -235,30 +326,28 @@ private void should_never_update_statement_id_for_conditional_updates(Session se
+      assertThat(row.isNull("d")).isTrue();
+      assertThat(ps.getPreparedId().resultSetMetadata.variables).isNull();
+      assertThat(ps.getPreparedId().resultSetMetadata.id).isEqualTo(idBefore);
+    }
```
```suggestion
    if (!isScylla) {
      assertThat(row.isNull("d")).isTrue();
      assertThat(ps.getPreparedId().resultSetMetadata.variables).isNull();
      assertThat(ps.getPreparedId().resultSetMetadata.id).isEqualTo(idBefore);
    } else if (supportsUseMetadataId) {
      assertThat(row.isNull("d")).isTrue();
      assertThat(ps.getPreparedId().resultSetMetadata.variables).hasSize(5);
      assertThat(ps.getPreparedId().resultSetMetadata.id).isNotEqualTo(idBefore);
    } else {
      assertThat(row.getColumnDefinitions()).doesNotContainVariable("d");
      assertThat(ps.getPreparedId().resultSetMetadata.variables).hasSize(4);
    }
```

nikagra commented on driver-core/src/main/java/com/datastax/driver/core/Host.java:
```diff
@@ -80,15 +73,17 @@ enum State {
   final ReentrantLock notificationsLock = new ReentrantLock(true);
 
   final ConvictionPolicy convictionPolicy;
+
```
My Git blame does not show this change being introduced by me

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Message.java:
```diff
@@ -350,10 +363,10 @@ EnumSet<Frame.Header.Flag> computeFlags(Request request) {
     }
 
-    int encodedSize(Request request) {
+    int encodedSize(Request request, ProtocolFeatureStore featureStore) {
```
It is already have ProtocolFeatureStore instance in it, use that one.
```suggestion
    int encodedSize(Request request) {
```

dkropachev commented on driver-core/src/main/java/com/datastax/driver/core/Message.java:
```diff
@@ -362,7 +375,7 @@ int encodedSize(Request request) {
     }
 
-    void encode(Request request, ByteBuf destination) {
+    void encode(Request request, ByteBuf destination, ProtocolFeatureStore featureStore) {
```
Same here
```suggestion
    void encode(Request request, ByteBuf destination) {
```
